File: src/main/java/com/nuutrai/reactor/util/VecLoc.java
package com.nuutrai.reactor.util;

import org.bukkit.Bukkit;
import org.bukkit.Location;
import org.bukkit.World;
import org.bukkit.util.Vector;
import org.jetbrains.annotations.NotNull;

import java.io.Serializable;
import java.util.ArrayList;
import java.util.List;
import java.util.Objects;
import java.util.UUID;

import static com.nuutrai.reactor.Reactor.logger;

public class VecLoc implements Serializable {

    private int x;
    private int y;
    private int z;
    private UUID world;

    public VecLoc(int x, int y, int z, UUID world) {
        this.world = world;
        this.z = z;
        this.y = y;
        this.x = x;
    }

    public VecLoc(Location location, UUID world) {
        this.x = location.getBlockX();
        this.y = location.getBlockY();
        this.z = location.getBlockZ();
        this.world = world;
    }

    public Location toLocation() {
        try {
            return new Location(getWorld(), x, y, z);
        } catch (NullPointerException e) {
            logger.severe("BUILD FALLBACK DUMB DUMB");
        }
        return new Location(Bukkit.getWorlds().getFirst(), x, y, z);
    }

    public Location toLocation(UUID world) {
        return new Location(getWorld(world), x, y, z);
    }

    public void setX(int x) {
        this.x = x;
    }

    public void setY(int y) {
        this.y = y;
    }

    public void setZ(int z) {
        this.z = z;
    }

    public void setWorld(UUID world) {
        this.world = world;
    }

    public World getWorld() {
        return getWorld(world);
    }

    public UUID getUUID() {
        return world;
    }

    public List<VecLoc> getNeighbours() {
        ArrayList<VecLoc> locs = new ArrayList<>();

        for (int i = 0; i < 4; i++) {
            locs.add(
                    switch (i) {
                        case 0: yield new VecLoc(x, y, z+1, world);
                        case 1: yield new VecLoc(x, y, z-1, world);
                        case 2: yield new VecLoc(x+1, y, z, world);
                        case 3: yield new VecLoc(x-1, y, z, world);
                        default:
                            throw new IllegalStateException("Unexpected value: " + i);
                    }
            );
        }

        return locs;

    }

    public static World getWorld(UUID world) {
        return Bukkit.getWorld(world.toString());
    }

    @Override
    public boolean equals(Object object) {
        if (this == object) return true;
        if (object == null || getClass() != object.getClass()) return false;
        VecLoc vecLoc = (VecLoc) object;
        return x == vecLoc.x && y == vecLoc.y && z == vecLoc.z && Objects.equals(world, vecLoc.world);
    }

//    public String getId() {
//        return String.valueOf(hashCode());
//    }

    public int getX() {
        return x;
    }

    public int getY() {
        return y;
    }

    public int getZ() {
        return z;
    }

    public int hashCode() {
        return Objects.hash(x, y, z, world);
    }
}

File: src/main/kotlin/com/nuutrai/reactor/util/VecLoc.kt
package com.nuutrai.reactor.util

import com.nuutrai.reactor.Reactor
import org.bukkit.Bukkit
import org.bukkit.Location
import org.bukkit.World
import java.io.Serializable
import java.util.*

class VecLoc : Serializable {
	//    public String getId() {
	//        return String.valueOf(hashCode());
	//    }
    var x: Int
    var y: Int
    var z: Int
	var uUID: UUID
		private set

	constructor(x: Int, y: Int, z: Int, world: UUID) {
		this.uUID = world
		this.z = z
		this.y = y
		this.x = x
	}

	constructor(location: Location, world: UUID) {
		this.x = location.blockX
		this.y = location.blockY
		this.z = location.blockZ
		this.uUID = world
	}

	fun toLocation(): Location {
		try {
			return Location(getWorld(), x.toDouble(), y.toDouble(), z.toDouble())
		} catch (_: NullPointerException) {
			Reactor.Companion.logger.severe("BUILD FALLBACK DUMB DUMB")
		}
		return Location(Bukkit.getWorlds()[0], x.toDouble(), y.toDouble(), z.toDouble())
	}

	fun toLocation(world: UUID): Location {
		return Location(getWorld(world), x.toDouble(), y.toDouble(), z.toDouble())
	}

	fun setWorld(world: UUID) {
		this.uUID = world
	}

	fun getWorld(): World? {
		return getWorld(this.uUID)
	}

	val neighbours: MutableList<VecLoc>
		get() {
			val locs = ArrayList<VecLoc>()

			locs.add(VecLoc(x + 1, y, z, this.uUID))
			locs.add(VecLoc(x - 1, y, z, this.uUID))
			locs.add(VecLoc(x, y, z + 1, this.uUID))
			locs.add(VecLoc(x, y, z - 1, this.uUID))

			return locs
		}

	override fun equals(other: Any?): Boolean {
		if (this === other) return true
		if (other == null || javaClass != other.javaClass) return false
		val vecLoc = other as VecLoc
		return x == vecLoc.x && y == vecLoc.y && z == vecLoc.z && this.uUID == vecLoc.uUID
	}

	override fun hashCode(): Int {
		return Objects.hash(x, y, z, this.uUID)
	}

	override fun toString(): String {
		return String.format("VecLoc@%s,%s,%s", this.x, this.y, this.z)
	}

	companion object {
		fun getWorld(world: UUID): World? {
			return Bukkit.getWorld(world.toString())
		}
	}
}
